v2/handler.js

```diff
@@ -0,0 +1,234 @@
+        return;
+    }
+    var id = resFrame.id;
+    var code = resFrame.body.code;
+    var arg1 = resFrame.body.arg1;
+    var arg2 = resFrame.body.arg2;
+    var arg3 = resFrame.body.arg3;
+    if (code === v2.CallResponse.Codes.OK) {
+        self.completeOutOp(null, id, arg2, arg3);
+    } else {
+        self.completeOutOp(TChannelApplicationError({
+            code: code,
+            arg1: arg1,
+            arg2: arg2,
+            arg3: arg3
+        }), id, arg2, null);
+    }
+};
+
+TChannelV2Handler.prototype.handleError = function handleError(errFrame) {
+    var self = this;
+    var id = errFrame.id;
+    var code = errFrame.body.code;
+    var message = errFrame.body.message;
+    var err = v2.ErrorResponse.CodeErrors[code]({message: message});
+    self.completeOutOp(err, id, null, null);
+};
+
+TChannelV2Handler.prototype.sendInitRequest = function sendInitRequest() {
+    var self = this;
+    var id = self.nextFrameId(); // TODO: assert(id === 1)?
+    var hostPort = self.channel.hostPort;
+    var processName = self.channel.processName;
+    var body = v2.InitRequest(v2.VERSION, hostPort, processName);
+    var reqFrame = v2.Frame(id, body);
+    self.writeFrame(reqFrame);
+};
+
+TChannelV2Handler.prototype.sendInitResponse = function sendInitResponse(reqFrame) {
+    var self = this;
+    var id = reqFrame.id;
+    var hostPort = self.channel.hostPort;
+    var processName = self.channel.processName;
+    var body = v2.InitResponse(v2.VERSION, hostPort, processName);
+    var resFrame = v2.Frame(id, body);
+    self.writeFrame(resFrame);
+};
+
+/* jshint maxparams:6 */
+TChannelV2Handler.prototype.sendRequestFrame = function sendRequestFrame(options, arg1, arg2, arg3, callback) {
+    var self = this;
+    var id = self.nextFrameId();
+    var flags = 0; // TODO: streaming
+    var ttl = options.timeout || 1; // TODO: better default, support for dynamic
+    var tracing = options.tracing || null; // TODO: generate
+    var service = options.service || null; // TODO: what even is this? default it to some sort of "my" name?
+    var headers = options.headers || {};
+    var csum;
+    if (options.checksum === undefined || options.checksum === null) {
+        csum = v2.Checksum.Types.FarmHash32;
+    } else {
+        csum = options.checksum;
+    }
+    var reqBody = v2.CallRequest(flags, ttl, tracing, service, headers, csum, arg1, arg2, arg3);
+    var reqFrame = v2.Frame(id, reqBody);
+    self.writeFrame(reqFrame, callback);
+    return id;
+};
+
+TChannelV2Handler.prototype.sendResponseFrame = function sendResponseFrame(reqFrame, headers, err, res1, res2, callback) {
+    // TODO: refactor this all the way back out through the op handler calling convention
+    var self = this;
+    var id = reqFrame.id;
+    var flags = 0; // TODO: streaming
+    var arg1 = reqFrame.body.arg1;
+    var checksumType = reqFrame.body.csum.type;
```

Comment:
does the checksum have to be the same as the one from the incoming request.

What if we always want to send farmhash back on the res because #speed